File: lingting-security-grpc/src/main/java/live/lingting/framework/endpoint/SecurityGrpcAuthorizationEndpoint.java
package live.lingting.framework.endpoint;

import com.google.protobuf.Empty;
import io.grpc.stub.StreamObserver;
import live.lingting.framework.convert.SecurityGrpcConvert;
import live.lingting.framework.security.annotation.Authorize;
import live.lingting.framework.security.authorize.SecurityAuthorizationService;
import live.lingting.framework.security.domain.AuthorizationVO;
import live.lingting.framework.security.domain.SecurityScope;
import live.lingting.framework.security.exception.AuthorizationException;
import live.lingting.framework.security.password.SecurityPassword;
import live.lingting.framework.security.resource.SecurityHolder;
import live.lingting.framework.security.store.SecurityStore;
import live.lingting.protobuf.SecurityGrpcAuthorization;
import live.lingting.protobuf.SecurityGrpcAuthorizationServiceGrpc;
import lombok.RequiredArgsConstructor;

/**
 * @author lingting 2023-12-18 15:31
 */
@RequiredArgsConstructor
public class SecurityGrpcAuthorizationEndpoint
	extends SecurityGrpcAuthorizationServiceGrpc.SecurityGrpcAuthorizationServiceImplBase {

	private final SecurityAuthorizationService service;

	private final SecurityStore store;

	private final SecurityPassword securityPassword;

	private final SecurityGrpcConvert convert;

	@Override
	public void logout(Empty request, StreamObserver<SecurityGrpcAuthorization.AuthorizationVO> observer) {
		SecurityScope scope = SecurityHolder.scope();
		store.deleted(scope);
		onNext(scope, observer);
	}

	@Override
	@Authorize(anyone = true)
	public void password(SecurityGrpcAuthorization.AuthorizationPasswordPO po,
						 StreamObserver<SecurityGrpcAuthorization.AuthorizationVO> observer) {
		String username = po.getUsername();
		String rawPassword = po.getPassword();
		String password = securityPassword.decodeFront(rawPassword);
		SecurityScope rawScope = service.validAndBuildScope(username, password);
		if (rawScope == null) {
			throw new AuthorizationException("Username or password is incorrect!");
		}
		SecurityScope scope = convert.scopeExpand(rawScope);
		store.save(scope);
		onNext(scope, observer);
	}

	@Override
	public void refresh(Empty request, StreamObserver<SecurityGrpcAuthorization.AuthorizationVO> observer) {
		SecurityScope scope = service.refresh(SecurityHolder.token());
		if (scope == null) {
			throw new AuthorizationException("Authorization has expired!");
		}
		store.update(scope);
		onNext(scope, observer);
	}

	@Override
	public void resolve(Empty request, StreamObserver<SecurityGrpcAuthorization.AuthorizationVO> observer) {
		SecurityScope scope = SecurityHolder.scope();
		onNext(scope, observer);
	}

	protected void onNext(SecurityScope scope, StreamObserver<SecurityGrpcAuthorization.AuthorizationVO> observer) {
		AuthorizationVO vo = convert.scopeToVo(scope);
		onNext(vo, observer);
	}

	protected void onNext(AuthorizationVO vo, StreamObserver<SecurityGrpcAuthorization.AuthorizationVO> observer) {
		SecurityGrpcAuthorization.AuthorizationVO authorizationVO = convert.toProtobuf(vo);
		observer.onNext(authorizationVO);
		observer.onCompleted();
	}

}

File: lingting-security-grpc/src/main/kotlin/live/lingting/framework/endpoint/SecurityGrpcAuthorizationEndpoint.kt
package live.lingting.framework.endpoint

import com.google.protobuf.Empty
import io.grpc.stub.StreamObserver
import live.lingting.framework.convert.SecurityGrpcConvert
import live.lingting.framework.protobuf.SecurityGrpcAuthorization
import live.lingting.framework.protobuf.SecurityGrpcAuthorizationServiceGrpc.SecurityGrpcAuthorizationServiceImplBase
import live.lingting.framework.security.SecurityEndpointService
import live.lingting.framework.security.annotation.Authorize
import live.lingting.framework.security.domain.AuthorizationVO
import live.lingting.framework.security.po.EndpointPasswordPO
import live.lingting.framework.security.po.EndpointTokenPO

/**
 * @author lingting 2023-12-18 15:31
 */
class SecurityGrpcAuthorizationEndpoint(
    val service: SecurityEndpointService,
    val convert: SecurityGrpcConvert,
) : SecurityGrpcAuthorizationServiceImplBase() {
    override fun logout(request: Empty, observer: StreamObserver<SecurityGrpcAuthorization.AuthorizationVO>) {
        val logout = service.logout()
        onNext(logout, observer)
    }

    @Authorize(anyone = true)
    override fun password(
        po: SecurityGrpcAuthorization.AuthorizationPasswordPO,
        observer: StreamObserver<SecurityGrpcAuthorization.AuthorizationVO>
    ) {
        val p = EndpointPasswordPO().apply {
            username = po.username
            password = po.password
        }
        val vo = service.password(p)
        onNext(vo, observer)
    }


    override fun refresh(request: SecurityGrpcAuthorization.TokenPO, observer: StreamObserver<SecurityGrpcAuthorization.AuthorizationVO>) {
        val p = EndpointTokenPO().apply {
            raw = request.raw
            type = request.type
            value = request.value
        }
        val vo = service.refresh(p)
        onNext(vo, observer)
    }

    @Authorize(anyone = true)
    override fun resolve(request: SecurityGrpcAuthorization.TokenPO, observer: StreamObserver<SecurityGrpcAuthorization.AuthorizationVO>) {
        val p = EndpointTokenPO().apply {
            raw = request.raw
            type = request.type
            value = request.value
        }
        val vo = service.resolve(p)
        onNext(vo, observer)
    }

    fun onNext(vo: AuthorizationVO, observer: StreamObserver<SecurityGrpcAuthorization.AuthorizationVO>) {
        val authorizationVO = convert.toProtobuf(vo)
        observer.onNext(authorizationVO)
        observer.onCompleted()
    }

}
